### ISSUE TITLE
Computing distances

### ISSUE THREAD

**Issue by videlec** [flatsurf/sage-flatsurf#293]
It is desirable to implement distances for points in a Euclidean or hyperbolic surface.

- [J. S. B. Mitchell, D. M. Mount and C. H. Papadimitriou "The Discrete Geodesic Problem" (1987)](https://epubs.siam.org/doi/10.1137/0216045)

**Comment by saraedum** [comment]
This is implemented in #213.

**Comment by saraedum** [comment]
> It is desirable to implement distances for points in a Euclidean or hyperbolic surface.
> 
>     * [J. S. B. Mitchell, D. M. Mount and C. H. Papadimitriou "The Discrete Geodesic Problem" (1987)](https://epubs.siam.org/doi/10.1137/0216045)

Their algorithm runs in $O(E^2 log E)$, however, that algorithm only applies to polyhedra. I don't see how the idea can be rescued for say translation surfaces. There are statements such as Lemma 4.1 which I think are crucial to the construction.

> An optimal path $p(x)$ to a point $x\in e$ [$e$ is an edge] passes through the interior of at most one face containing $x$.

The implementation that I put in before you had posted that reference is somewhat naive and runs in $O(V^4)$ (maybe somehow $O(V^3)$ with better analysis of the runtime).

**Comment by videlec** [comment]
Indeed, the statements in the article looks very wrong! If the surface is "reasonable" but the triangulation is "crazy", then there is no way you can achieve $O(E^2 log E)$.

**Comment by saraedum** [comment]
So, the only viable approach seems to be to reduce the problem to a saddle connection search with good upper bounds. This is what I did in #213 which is a lot of Floyd which is why I said initially $O(V^4)$ but actually it has a saddle connection search in there and that cannot be immediately bounded by something that only depends on the combinatorics of the surface.

**Comment by saraedum** [comment]
On a much more elementary level, the length of a shortest path is a sum of square roots. It's annoying to even compare the lengths of two paths…

**Comment by saraedum** [comment]
Also, we do not have segments in sage-flatsurf yet. Only the StraightLineTrajectory, so it's a bit hard to represent such a shortest path. I'd say that essentially each segment should be a tangent vector + holonomy.

**Comment by videlec** [comment]
> On a much more elementary level, the length of a shortest path is a sum of square roots. It's annoying to even compare the lengths of two paths…

I think it would be good to switch from e-antic to flint via https://github.com/flintlib/flint/issues/1848 (though this is lacking the C++ interface). Then one could use one of the two Frederik implementations of AA.
